DOC Add CorrelationRemover user guide (#1029)
fairlearn/fairlearn: docs/user_guide/mitigation.rst

   
.. currentmodule:: fairlearn.preprocessing

Preprocessing algorithms transform the dataset to mitigate possible unfairness
present in the data.
Preprocessing algorithms in Fairlearn follow the :class:`sklearn.base.TransformerMixin`
class, meaning that they can :code:`fit` to the dataset and :code:`transform` it
(or :code:`fit_transform` to fit and transform in one go).

.. _correlation_remover:

Correlation Remover
~~~~~~~~~~~~~~~~~~~
Sensitive features can be correlated with non-sensitive features in the dataset.
By applying the :code:`CorrelationRemover`, these correlations are projected away

> whiledetails from the original data are retained as much as possible (as measured

Reviewer: ```suggestion
while details from the original data are retained as much as possible (as measured
```

by the least-squares error). The user can control the level of projection via the
:code:`alpha` parameter. In mathematical terms, assume we have the original dataset
:math:`X` which contains a set of sensitive attributes :math:`S` and a set of
non-sensitive attributes :math:`Z`. The removal of correlation is then

> mathematically described as:

Reviewer: I would use either 'mathematical term' OR 'mathematically described'  but not both.

.. math::
        \min _{\mathbf{z}_{1}, \ldots, \mathbf{z}_{n}} \sum_{i=1}^{n}\left\|\mathbf{z}_{i}
        -\mathbf{x}_{i}\right\|^{2} \\
        \text{subject to} \\
        \frac{1}{n} \sum_{i=1}^{n} \mathbf{z}_{i}\left(\mathbf{s}_{i}-\overline{\mathbf{s}}
        \right)^{T}=\mathbf{0}

The solution to this problem is found by centering sensitive features, fitting a
linear regression model to the non-sensitive features and reporting the residual.

> The columns in :math:`S` will be dropped from the dataset :math:`X` but the

Reviewer: ```suggestion
The columns in :math:`S` will be dropped from the dataset :math:`X`.
```

> amount of correlation that is removed can be controlled using the

Reviewer: ```suggestion
The amount of correlation that is removed can be controlled using the
```

:code:`alpha` parameter. This is described as follows:

.. math::
        X_{\text{tfm}} = \alpha X_{\text{filtered}} + (1-\alpha) X_{\text{orig}}

> Note that the lack of correlation does not imply anything about statistical dependence.

Reviewer: Maybe add an additional line to explain that correlation measures linear relationships and that non-linear dependencies can still exist in the dataset.

original dataset. We are particularly interested in the correlations
between the 'sex_Female' column and the three non-sensitive attributes
'age', 'fnlwgt' and 'education-num'. The target variable is also
included in these visualization for completeness, and it is defined as
a binary feature which indicated an income higher than $50k. We see that
'sex_Female' is not highly correlated with the three mentioned attributes,
but we want to remove these correlations nonetheless.
The code for generating the correlation matrix can be found in
`this example notebook
<../auto_examples/plot_correlationremover_before_after.html>`_.

.. figure:: ../auto_examples/images/sphx_glr_plot_correlationremover_before_after_001.png
    :align: center
    :target: ../auto_examples/plot_correlationremover_before_after.html

In order to see the effect of :class:`CorrelationRemover`, we visualize
how the correlation matrix has changed after the transformation of the
dataset. Due to rounding, some of the 0.0 values appear as -0.0. Either
way, the :code:`CorrelationRemover` successfully removed all correlation
between 'sex_Female' and the other columns while retaining the correlation

> between the other features.

Reviewer: Maybe note that the correlation of age with the target variable has decreased a bit as well.

Reviewer: Though we might want to wait with such details after we have replaced the Adult dataset in this example.

Author: That's what I thought as well :). I tried to keep the text as independent from the dataset as possible, although that is not always possible ofcourse.

        \text{subject to} \\
        \frac{1}{n} \sum_{i=1}^{n} \mathbf{z}_{i}\left(\mathbf{s}_{i}-\overline{\mathbf{s}}
        \right)^{T}=\mathbf{0}

The solution to this problem is found by centering sensitive features, fitting a
linear regression model to the non-sensitive features and reporting the residual.
The columns in :math:`S` will be dropped from the dataset :math:`X`.
The amount of correlation that is removed can be controlled using the
:code:`alpha` parameter. This is described as follows:

.. math::
        X_{\text{tfm}} = \alpha X_{\text{filtered}} + (1-\alpha) X_{\text{orig}}

Note that the lack of correlation does not imply anything about statistical dependence.
In particular, since correlation measures linear relationships, it might still be
possible that non-linear relationships exist in the data. Therefore, we expect this
to be most appropriate as a preprocessing step for (generalized) linear models.

In the example below, the `Diabetes 130-Hospitals <https://www.openml.org/d/43874>`_
is loaded and the correlation between the African American race and
the non-sensitive features is removed. This dataset contains more races,

> but for the cause of this example we will only focus on the African

Reviewer: ```suggestion
but in this example we will only focus on the African
```

> American race. This procedure will drop the sensitive features from the dataset.

Reviewer: Nitpick: on a first read it's not immediately clear whether "this procedure" related to manual pre-processing or CorrelationRemover

   
.. currentmodule:: fairlearn.preprocessing

Preprocessing algortihms transform the dataset to mitigate possible unfairness
present in the data.
Preprocessing algorithms in Fairlearn follow the `scikit-learn transformers
<https://scikit-learn.org/stable/modules/generated/sklearn.base.TransformerMixin.html#sklearn.base.TransformerMixin>`_
class, meaning that they can :code:`fit` to the dataset and :code:`transform` it.

> A :code:`fit_transform` might not always be available, in which case the separate

Reviewer 2: I think that `fit_transform` should be provided automatically by `TransformerMixin`, please double check whether `CorrelationRemover` supports it (I think it might?).

> A :code:`fit_transform` might not always be available, in which case the separate

Author: You are correct! `CorrelationRemoer` supports `fit_transform`, I overlooked this, thanks :)

:code:`fit` and :code:`transform` methods need to be used in sequence.

.. _correlation_remover:

Correlation Remover
~~~~~~~~~~~~~~~~~~~
Sensitive features can be correlated with non sensitive features in the dataset.
By applying the :code:`CorrelationRemover`, these correlations are filtered out while

> as much as possible information is retained (as measured by the least-squares error).

Reviewer 3: I think we should either mention the `alpha` parameter here or include something about how you can control the level of "filtering" of correlation, so users understand it's not always complete removal of correlation.

Author: I agree with both these comments! I've mentioned the `alpha` parameter earler in the doc now, as well as using a different wording for "information".

This is mathemetically described as:

.. math::
        \min _{\mathbf{z}_{1}, \ldots, \mathbf{z}_{n}} \sum_{i=1}^{n}\left\|\mathbf{z}_{i}
        -\mathbf{x}_{i}\right\|^{2} \\
        \text{subject to} \\
        \frac{1}{n} \sum_{i=1}^{n} \mathbf{z}_{i}\left(\mathbf{s}_{i}-\overline{\mathbf{s}}
        \right)^{T}=\mathbf{0}

> In the example below, a subset of the `adult dataset <https://www.openml.org/d/1590>`_

Reviewer 4: I'm not a huge fan of introducing another example with adult... I have a PR where one of the main blockers was that I am using adult. I feel that this should be a blocker for all PRs that introduce it.

Author: I can understand that! Is there an alternative dataset we are also fine with using, or is that still up for debate?

Reviewer 4: It's one of the discussions that keep resurfacing... I'm afraid I don't have a great answer to this. Perhaps worth raising in the community call unless it's already been discussed there(?) @fairlearn/fairlearn-maintainers

Reviewer 2: You should get unblocked once #1064 is resolved. Diabetes 130-hospitals should be more suitable.

is loaded and the correlation between sex and the non-sensitive features is removed.

.. doctest:: mitigation
    :options:  +NORMALIZE_WHITESPACE

    >>> from fairlearn.preprocessing import CorrelationRemover
    >>> import pandas as pd
    >>> from sklearn.datasets import fetch_openml
    >>> data = fetch_openml(data_id=1590, as_frame=True)
    >>> X = data.data[['age', 'fnlwgt', 'education-num', 'sex']]
    >>> X = pd.get_dummies(X)
    >>> cr = CorrelationRemover(sensitive_feature_ids=['sex_Female', 'sex_Male'])
    >>> cr.fit(X)
    CorrelationRemover(sensitive_feature_ids=['sex_Female', 'sex_Male'])
    >>> X_transform = cr.transform(X)

> ..  I am not sure whether it would be valuable to show the transformed data here,

Reviewer 4: What do you have in mind in terms of showing?

Author: I had in mind to just display the transformed values, such that the user gets an i dea of what it would look like, e.g. the below screenshot. However, maybe it is not really valuable to show that to the user, hence my comment in the code :).
![Transformed data example](https://user-images.githubusercontent.com/33757199/156995915-bfed736c-fe31-4bbe-9629-dd78f397b36a.png)

Author: How about the following? I recall from some papers and/or lecetures that t-SNE is quite nice for these kind of visualizations, and I think it looks quite good and intuitive. I also tried PCA, but that didn't look quite as nice as the t-SNE plots.
![CR](https://user-images.githubusercontent.com/33757199/160357446-88a06edf-c94f-40c7-8ffc-cffd9d846390.png)
![No CR](https://user-images.githubusercontent.com/33757199/160357448-1f5e4893-1c78-4100-a9cb-537e47095b1d.png)

Author: I tried both PCA and t-SNE, but both of them showed very little difference in the before and after plots. I figured it would perhaps be most intuitive to show the correlation matrices before and after running the `CorrelationRemover`, so that is what I have done right now by making an example notebook and including the figures in the user guide :)

Reviewer 4: Ah yes t-SNE is what I was thinking if without remembering the name, thanks! Hmm, that makes it tricky...

I agree with your process and conclusion. If any other @fairlearn/fairlearn-maintainers or others have thoughts on what might be useful to add please chime in! I guess the most important part is always how it changes outcomes, but it's really interesting how it removes correlations, too, hence a correlation matrix makes sense. 👍🏻 Might be worth pointing out the important entries, though, because people (or is that just me) don't read matrices 😄

Reviewer: My suggestion would be to show a heatmap (with colors as well as numbers) of the *differences* in correlations before/after (e.g., a value of 0.2 would mean that correlation between X1 and X2 has increased with 0.2). That allows you to see which correlations were affected the most (and in what way) in a single glance.

   
.. currentmodule:: fairlearn.preprocessing

Preprocessing algorithms transform the dataset to mitigate possible unfairness
present in the data.
Preprocessing algorithms in Fairlearn follow the `scikit-learn transformers
<https://scikit-learn.org/stable/modules/generated/sklearn.base.TransformerMixin.html#sklearn.base.TransformerMixin>`_
class, meaning that they can :code:`fit` to the dataset and :code:`transform` it
(or :code:`fit_transform` to fit and transform in one go).

.. _correlation_remover:

Correlation Remover
~~~~~~~~~~~~~~~~~~~
Sensitive features can be correlated with non-sensitive features in the dataset.

> By applying the :code:`CorrelationRemover`, these correlations are filtered out while

Reviewer: I am not super happy with the word "filtering" here, but haven't been able to come up with a good alternative... Thoughts? @romanlutz @rensoostenbach @LeJit

Reviewer 4: Projected away? 😄 
@koaning may have thoughts as he came up with this 🙂

Reviewer 5: "projected away" sounds fine 👍

details from the original data are retained as much as possible (as measured by the
least-squares error). The user can control the level of filtering via the :code:`alpha`
parameter. The removal of correlation is mathematically described as:

.. math::
        \min _{\mathbf{z}_{1}, \ldots, \mathbf{z}_{n}} \sum_{i=1}^{n}\left\|\mathbf{z}_{i}
        -\mathbf{x}_{i}\right\|^{2} \\
        \text{subject to} \\
        \frac{1}{n} \sum_{i=1}^{n} \mathbf{z}_{i}\left(\mathbf{s}_{i}-\overline{\mathbf{s}}
        \right)^{T}=\mathbf{0}

The amount of filtering (as controlled with the :code:`alpha` parameter) is then
described as follows:

.. math::
        X_{\mathrm{tfm}}=\alpha X_{\text {filtered }}+(1-\alpha) X_{\mathrm{orig}}

In the example below, a subset of the `adult dataset <https://www.openml.org/d/1590>`_
is loaded and the correlation between sex and the non-sensitive features is removed.
This procedure will drop the sensitive features from the dataset.

> 

Reviewer: Perhaps the narrative could be improved by first visualizing the correlations in the original plot ("look there are features correlated with the sensitive attribute!"), after which we apply `CorrelationRemover` and show how the correlations have changed (with the plot you showed before). 

In my view it would also be useful to include the target variable in these correlation matrixes.

.. doctest:: mitigation
    :options:  +NORMALIZE_WHITESPACE

    >>> from fairlearn.preprocessing import CorrelationRemover
    >>> import pandas as pd
    >>> from sklearn.datasets import fetch_openml
    >>> data = fetch_openml(data_id=1590, as_frame=True)
    >>> X = data.data[['age', 'fnlwgt', 'education-num', 'sex']]
    >>> X = pd.get_dummies(X)
    >>> X = X.drop('sex_Male', axis=1)
    >>> cr = CorrelationRemover(sensitive_feature_ids=['sex_Female'])
    >>> cr.fit(X)
    CorrelationRemover(sensitive_feature_ids=['sex_Female'])
    >>> X_transform = cr.transform(X)

> In order to see what the :code:`CorrelationRemover` has done,

Reviewer: How about something like:

In order to see the effect of  :class:`CorrelationRemover`, we visualize how the correlation matrix has changed after the transformation of the dataset.

we plot the correlation matrix of this dataset. In fact, we plot
the differences in correlation values before and after the
:code:`CorrelationRemover` was applied. Below those values, the
percentual change in correlation is shown. Due to rounding, some of the
0.0 values appear as -0.0. Either way, the :code:`CorrelationRemover`
successfully removed all correlation between 'sex_Female' and the other
columns while retaining the correlation between the other features.
The code for generating these matrices can be found
in `this example notebook
<../auto_examples/plot_correlationremover_before_after.html>`_.

.. figure:: ../auto_examples/images/sphx_glr_plot_correlationremover_before_after_001.png
    :align: center
    :target: ../auto_examples/plot_correlationremover_before_after.html

> We can also use the :code:`alpha` parameter with for instance :math:`\alpha=0.5`

Reviewer: I would expect a second visualization here.

Note that the lack of correlation does not imply anything about statistical dependence.
Therefore, we expect this to be most appropriate as a preprocessing step for

> (generalized) linear models. Furthermore, since correlation measures linear

Reviewer: I would move the "Furthermore, since correlation [...] in the data." *before* "Therefore, we expect [...] linear models." and replace "Furthermore" by "In particular"
